sky/engine/platform/fonts/apple/FontCacheIOS.mm

```diff
@@ -0,0 +1,178 @@
+#include "sky/engine/platform/fonts/FontCache.h"
+#include "base/mac/scoped_nsobject.h"
+#include "base/lazy_instance.h"
+#include "base/macros.h"
+#include "base/logging.h"
+
+namespace blink {
+
+template <class T>
+class CFRef {
+ public:
+  CFRef() : _ref(nullptr) {}
+
+  CFRef(T t) : _ref(t) {}
+
+  CFRef(const CFRef&) = delete;
+
+  void operator=(const CFRef&) = delete;
+
+  CFRef(CFRef&& o) : _ref(o._ref) { o._ref = nullptr; }
+
+  void operator=(CFRef&& o) {
+    _ref = o._ref;
+    o._ref = nullptr;
+  }
+
+  operator bool() const { return _ref != nullptr; }
+
+  T get() const { return _ref; }
+
+  ~CFRef() {
+    if (_ref != nullptr) {
+      CFRelease(reinterpret_cast<CFTypeRef>(_ref));
+    }
+  }
+
+ private:
+  T _ref;
+};
+
+class FontFallbackSelector {
+ public:
+  FontFallbackSelector() {
+    UIFontDescriptor* ui_desc =
+        [UIFont systemFontOfSize:[UIFont systemFontSize]].fontDescriptor;
+
+    CFRef<CTFontDescriptorRef> desc(CTFontDescriptorCreateWithNameAndSize(
+        reinterpret_cast<CFStringRef>(ui_desc.postscriptName),
+        ui_desc.pointSize));
+
+    if (!desc) {
+      return;
+    }
+
+    _prototype = CFRef<CTFontRef>{
+        CTFontCreateWithFontDescriptor(desc.get(), ui_desc.pointSize, nullptr)};
+  }
+
+  CFRef<CTFontDescriptorRef> fallbackFont(unichar glyph) {
+    if (!_prototype) {
+      return {};
+    }
+
+    CFRef<CFStringRef> unicode_string(CFStringCreateWithFormat(
+        kCFAllocatorDefault, nullptr, CFSTR("%C"), glyph));
+
+    if (!unicode_string) {
+      return {};
+    }
+
+    CFRef<CTFontRef> font(CTFontCreateForString(
+        _prototype.get(), unicode_string.get(),
+        CFRangeMake(0, CFStringGetLength(unicode_string.get()))));
+
+    if (!font) {
+      return {};
+    }
+
+    return CTFontCopyFontDescriptor(font.get());
+  }
+
+ private:
+  CFRef<CTFontRef> _prototype;
+
+  DISALLOW_COPY_AND_ASSIGN(FontFallbackSelector);
+};
+
+static base::LazyInstance<FontFallbackSelector> g_fallback_selector =
+    LAZY_INSTANCE_INITIALIZER;
+
+void FontCache::getFontForCharacter(
+    UChar32 c,
+    const char*,
+    FontCache::PlatformFallbackFont* fallbackFont) {
+  if (fallbackFont == nullptr) {
+    return;
+  }
+
+  fallbackFont->name = "";
+  fallbackFont->filename = "";
+  fallbackFont->ttcIndex = 0;
+  fallbackFont->fontconfigInterfaceId = 0;
+  fallbackFont->isBold = false;
+  fallbackFont->isItalic = false;
+
+  CFRef<CTFontDescriptorRef> font =
+      g_fallback_selector.Pointer()->fallbackFont(c);
+
+  if (!font) {
+    return;
+  }
+
+  CFRef<CFURLRef> cf_url(reinterpret_cast<CFURLRef>(
+      CTFontDescriptorCopyAttribute(font.get(), kCTFontURLAttribute)));
+
+  if (!cf_url) {
+    return;
+  }
+
+  CFRef<CFURLRef> cf_absolute_url(CFURLCopyAbsoluteURL(cf_url.get()));
+
+  if (!cf_absolute_url) {
+    return;
+  }
+
+  const NSURL* url =
+      reinterpret_cast<const NSURL*>(cf_absolute_url.get());
+
+  fallbackFont->filename = [url fileSystemRepresentation];
```

Comment:
From Docs:

`This string is automatically freed in the same way that a returned object would be released. The caller must either copy the string or use...`

The `blink::CString` constructor should copy out the string before the ARP pops.